arch/x86/ectx.c

```diff
@@ -43,22 +43,89 @@
 #include <uk/hexdump.h>
 #include <string.h> /* memset */
 
-enum x86_save_method {
-	X86_SAVE_NONE = 0,
-	X86_SAVE_FSAVE,
-	X86_SAVE_FXSAVE,
-	X86_SAVE_XSAVE,
-	X86_SAVE_XSAVEOPT
-};
-
-static enum x86_save_method ectx_method;
-static __sz ectx_size;
 static __sz ectx_align = 0x0;
+static __sz ectx_size;
+
+static void ectx_store_none(struct ukarch_ectx *state __unused) { }
+
+static void ectx_load_none(struct ukarch_ectx *state __unused) { }
+
+static void ectx_sanitize_none(struct ukarch_ectx *state __unused) { }
+
+static void ectx_store_fsave(struct ukarch_ectx *state)
+{
+	__asm__ __volatile__("fsave (%0)" :: "r"(state) : "memory");
+}
+
+static void ectx_load_frstor(struct ukarch_ectx *state)
+{
+	__asm__ __volatile__("frstor (%0)" :: "r"(state));
+}
 
-static void _init_ectx_store(void)
+static void ectx_store_fxsave(struct ukarch_ectx *state)
+{
+	__asm__ __volatile__("fxsave (%0)" :: "r"(state) : "memory");
+}
+
+static void ectx_load_fxrstor(struct ukarch_ectx *state)
+{
+	__asm__ __volatile__("fxrstor (%0)" :: "r"(state));
+}
+
+/* The specific state components restored correspond to the bits set in the
+ * requested-feature bitmap (RFBM), which is the logical-AND of EDX:EAX and
+ * XCR0. So, save everything ans use mask 0xffffffff:0xffffffff.
+ */
+static void ectx_store_xsave(struct ukarch_ectx *state)
+{
+	__asm__ __volatile__("xsave (%0)" :: "r"(state),
+			     "a"(0xffffffff), "d"(0xffffffff) : "memory");
+}
+
+static void ectx_store_xsaveopt(struct ukarch_ectx *state)
+{
+	__asm__ __volatile__("xsaveopt (%0)" :: "r"(state),
+			     "a"(0xffffffff), "d"(0xffffffff) : "memory");
+}
+
+static void ectx_load_xrstor(struct ukarch_ectx *state)
+{
+	__asm__ __volatile__("xrstor (%0)" :: "r"(state),
+			     "a"(0xffffffff), "d"(0xffffffff) : "memory");
+}
+
+static void ectx_sanitize_xsave(struct ukarch_ectx *state)
+{
+	/* XSAVE* & XRSTOR rely on sane values in the XSAVE header
+	 * (64 bytes starting at offset 512 from the base address)
+	 * and will raise #GP on garbage data. We must zero them out.
+	 */
+	((__u64 *)state)[64] = 0;
+	((__u64 *)state)[65] = 0;
+	((__u64 *)state)[66] = 0;
+	((__u64 *)state)[67] = 0;
+	((__u64 *)state)[68] = 0;
+	((__u64 *)state)[69] = 0;
+	((__u64 *)state)[70] = 0;
+	((__u64 *)state)[71] = 0;
+}
+
+typedef void (*ukarch_ectx_store_fn_t)(struct ukarch_ectx *);
+typedef void (*ukarch_ectx_load_fn_t)(struct ukarch_ectx *);
+typedef void (*ukarch_ectx_sanitize_fn_t)(struct ukarch_ectx *);
+
+ukarch_ectx_store_fn_t ectx_store_fn;
```

Comment:
> AFAICT (pls correct me if I'm wrong) this PR changes ectx handling overhead from
>
>direct func call (uk_ectx_*) + global mem read (ectx_method) + switch branching
>to
>direct func call (uk_ectx_*) + global mem read (ectx_*_fn) + indirect func call

yes. I want to be able to assert if assertions are enabled.

Is it wrong to assume that a fully optimized build without assertions would do just that (what you proposed)?

Reply:
> Is it wrong to assume that a fully optimized build without assertions would do just that (what you proposed)?

There's some chance that a good LTO pass _might_ decide to inline `uk_ectx_` calls (assuming the `|isr` thing doesn't negatively impact LTO, you need to check) so ectx calls end up 1 indirect call in the final binary, but that's not guaranteed. If we make it explicit in code it will be certain.

> yes. I want to be able to assert if assertions are enabled without copy-pasting them everywhere.

I think you would need far fewer asserts than right now if you directly export func ptrs, especially if you implement the other comment regarding init values:
- `ectx_align` (i.e., checking whether inited): should not be necessary since (1) a "real" ectx func can get called only after init (bc they're static and accessible only via func ptrs inited by you) and (2) the pre-init value of `ectx_*_none` is a no-op that doesn't care
- `state`: not strictly needed since it only crashes on null ptrs that would trigger a crash anyway :shrug: 
- `ectx_*_fn`: you _are_ that function, obviously it exists :smile: 
- `IS_ALIGNED(state, ectx_align)`: this one is the only that really is useful, and IMO it would be faster if it had a constant to check against instead of a runtime-assigned global var (LTO can't help you here I'm afraid). So there would be a tangible benefit to moving this assert into individual `uk_ectx_*` funcs.

Reply:
> ectx_align (i.e., checking whether inited): should not be necessary since (1) a "real" ectx func can get called only after init (bc they're static and accessible only via func ptrs inited by you) and (2) the pre-init value of ectx_*_none is a no-op that doesn't care

Actually, since there are no 64-bit machines that do not have at least SSE2, those none variants for store/load are useless I guess, I may as well initialize statically to FSAVE/FRSTOR.

> There's some chance that a good LTO pass might decide to inline uk_ectx_ calls (assuming the |isr thing doesn't negatively impact LTO, you need to check) so ectx calls end up 1 indirect call in the final binary, but that's not guaranteed. If we make it explicit in code it will be certain.

```
# compiler flags to prevent use of extended (FP, SSE, AVX) registers.
# This is for files that contain trap/exception/interrupt handlers
ISR_ARCHFLAGS += -mno-80387 -mno-mmx -mno-sse -mno-avx
ISR_ARCHFLAGS-$(call have_gcc)	+= -mno-fp-ret-in-387
ISR_ARCHFLAGS-$(call gcc_version_ge,7,1) += -mgeneral-regs-only
```
Hmm, I don't think it would be bad for LTO.

> state: not strictly needed since it only crashes on null ptrs that would trigger a crash anyway 🤷

That would be worse. I would rather go through the exception hoops of an explicit crash than the hoops of an unhandled trap which would cause  more confusion to a user for example that all they would see is a runtime crash instead of an assertion.

> I think you would need far fewer asserts 

I agree with this and the rest. The only reason I have this indirection is so that we keep assertions in one place as well as have one centralized function where we can place prints. However, I am still skeptical on whether it is worth making this function always a function pointer as I am pretty sure it will be optimized always. I am waiting for other's opinions as well here.

Reply:
> That would be worse. I would rather go through the exception hoops of an explicit crash than the hoops of an unhandled trap which would cause more confusion to a user for example that all they would see is a runtime crash instead of an assertion.

You'd need to do that anyway for all non-NULL invalid ptrs, so you're not saving yourself _that_ much fuss.

> The only reason I have this indirection is so that we keep assertions in one place as well as have one centralized function where we can place prints

Then perhaps defining `uk_ectx_*` as inlines (+ exporting the func ptrs & size & align) in `ctx.h` would net us the best of both worlds: centralised assertions & prints asd well as inlining without relying on LTO.

Reply:
> You'd need to do that anyway for all non-NULL invalid ptrs, so you're not saving yourself that much fuss.

In the code that I submit I assert every single argument as well as their fields (if structs) that I use. I find it good practice, has 0 runtime cost on production build and can save you time. Just my style/belief. :man_shrugging:

Reply:
> it'll probably also require small changes to ARM ectx code to bring in line

oh yes, that 100% will need to happen as ARM only has one functio for each... just NEON registers. Unlike the variable alignment/size/layout of x86 :zany_face: :dumpling:

Reply:
Hum, it is actually an interesting  idea to provide the interface with function pointers. In fact we are otherwise only replacing one jump of the switch case with an indirect function call with this pr.

Assertions could also be handled with surrounding function macros or static inline functions 🤔. Actually moving ukarch_ectx_load and co. to the header as static inline   might achieve the same thing and is still well readable source. Overall I would make sure that the individual implantations stay static.

Reply:
Ok then how about making `ukarch_ectx_load/store` as static inlines with the asserts and those static inlines call the function pointer? I will have to adapt ARM code as well.

Reply:
Yes, this is what I thought as a good variant to achieve better compiler optimization while still easy readable.